File: bridge/AudioScheduledSourceNode.cpp
#include "dart_api/dart_api.h"
#include "LabSound/LabSound.h"
#include "KeepNode.cpp"
using namespace lab;

DART_EXPORT int AudioScheduledSourceNode_isPlayingOrScheduled(int nodeId) {
    auto node = std::static_pointer_cast<AudioScheduledSourceNode>(getNode(nodeId));
    return node ? node->isPlayingOrScheduled() : 0;
}

DART_EXPORT void AudioScheduledSourceNode_stop(int nodeId, float when) {
    auto node = std::static_pointer_cast<AudioScheduledSourceNode>(getNode(nodeId));
    if(node) node->stop(when);
}

DART_EXPORT int AudioScheduledSourceNode_hasFinished(int nodeId) {
    auto node = std::static_pointer_cast<AudioScheduledSourceNode>(getNode(nodeId));
    return node ? node->hasFinished() : 0;
}

DART_EXPORT uint64_t AudioScheduledSourceNode_startWhen(int nodeId) {
    auto node = std::static_pointer_cast<AudioScheduledSourceNode>(getNode(nodeId));
    return node ? node->startWhen() : 0;
}

DART_EXPORT void AudioScheduledSourceNode_start(int nodeId, float when) {
    auto node = std::static_pointer_cast<AudioScheduledSourceNode>(getNode(nodeId));
    if(node) node->start(when);
}

DART_EXPORT int AudioScheduledSourceNode_playbackState(int nodeId) {
    auto node = std::static_pointer_cast<AudioScheduledSourceNode>(getNode(nodeId));
    return node ? static_cast<int>(node->playbackState()) : 0;
}

File: lab_sound_ffi/lib/src/lab_sound/audio_scheduled_source_node.dart
import 'audio_context.dart';
import 'audio_node.dart';
import 'lab_sound.dart';

class AudioScheduledSourceNode extends AudioNode {
  AudioScheduledSourceNode(AudioContext ctx, int nodeId): super(ctx, nodeId);
  bool get isPlayingOrScheduled => LabSound().AudioScheduledSourceNode_isPlayingOrScheduled(nodeId) > 0;
  stop({double when = 0}) => LabSound().AudioScheduledSourceNode_stop(nodeId, when);
  start({double when = 0}) => LabSound().AudioScheduledSourceNode_start(nodeId, when);
  bool get hasFinished => LabSound().AudioScheduledSourceNode_hasFinished(nodeId) > 0;
  int get startWhen => LabSound().AudioScheduledSourceNode_startWhen(nodeId);
  SchedulingState get playbackState => SchedulingState.values[LabSound().AudioScheduledSourceNode_playbackState(nodeId)];
}